doc/tutorials/debug.rst

```diff
@@ -0,0 +1,371 @@
+
+.. figure:: images/debug_image17.png
+   :align: center
+
+   int information
+
+cpuid
+=====
+
+"cupid cpu_num" provides the highest EAX call parameter (leaf)
+supported by the CPU in the EAX register and the values of EBX, ECX, and EDX
+registers.
+
+.. figure:: images/debug_image15.png
+   :align: center
+
+   cpuid information
+
+rdmsr/wrmsr
+===========
+
+We can read and write model specific register (MSR) to get and set register
+values through ``rdmsr/wrmsr register_address``.
+
+In the following example, we can get IA32_APIC_BASE value through
+the command::
+
+   rdmsr 1b
+
+and see that 1B (Hexadecimal) is the IA32_APIC_BASE MSR address.
+
+.. figure:: images/debug_image16.png
+   :align: center
+
+   IA32_APIC_BASE register information
+
+.. figure:: images/debug_image18.png
+   :align: center
+
+   rdmsr information
+
+An example
+==========
+
+As an example, we'll show how to obtain the interrupts of a pass-through USB device.
+
+First, we can get the USB controller BDF number (0:15.0) through the
+following command::
+
+   lspci | grep "USB controller"
+
+
+.. figure:: images/debug_image19.png
+   :align: center
+
+   USB controller BDF information
+
+Second, we use the ``pt`` command and use this BDF number
+to find the interrupt vector (VEC) "0x31".
+
+.. figure:: images/debug_image20.png
+   :align: center
+
+   USB controller interrupt information
+
+Finally we use the ``int`` command and use this interrupt vector
+to find the interrupt number (909) on CPU3.
+
+.. figure:: images/debug_image21.png
+   :align: center
+
+   USB controller interrupt number information
+
+ACRN Log
+********
+
+ACRN log provides console log and acrnlog for a user to analyze.
+We can use console log to debug directly, while acrnlog is a userland tool
+used to capture a ACRN hypervisor log.
+
+console log/acrn log
+====================
+
+We can set console log level and acrnlog level through::
+
+   loglevel [console_loglevel] [mem_loglevel]
+
+For example, we add the following code into function ``shell_cmd_help`` of
+the source file ``acrn-hypervisor/hypervisor/debug/shell.c``..
+
+.. figure:: images/debug_image22.png
+   :align: center
+
+   shell_cmd_help added information
+
+We set console log level to 5, and acrnlog level to 2 through the
+command::
+
+  loglevel 5 2
+
+Then we input "help" into ACRN console, and find the log as follows.
+
+.. figure:: images/debug_image23.png
+   :align: center
+
+   console log information
+
+Then we use the command::
+
+   sos_console
+
+to switch to the SOS console the use the commands::
+
+   systemctl enable acrnlog
+   reboot
+
+to enable the acrnlog and reboot the SOS. Next we set the console and
+acrnlog levels with::
+
+   loglevel 5 2
+
+and run the command::
+
+   help
+
+Finally we get into the SOS and use the command::
+
+   cat /tmp/acrnlog/acrnlog_cur.0
+
+on the console, and we'll see the following log:
+
+.. figure:: images/debug_image24.png
+   :align: center
+
+   acrnlog information
```

Comment:
OK, I think I understand now what you were getting at. It's rather confusing though. We need a proper paragraph that explains how to instrument the existing code *if* it's not sufficiently instrumented already. I would assume that the first step when debugging is really to turn on the logging info. If **and only** if it's not providing enough information, then we can guide the user on how to instrument the code. For this latter part, we should refer to these `pr_*`  functions and document them.